### PR TITLE
Improve the TLSv1.3 PSK implementation

### PR DESCRIPTION
**Pull request by mattcaswell**
This PR makes these changes:

1) Tolerate PSKs that are not the same size as the hash size

2) Default to SHA-256 as per the spec if we have a PSK that we don't know the hash for

3) If we don't have TLSv1.3 style PSK callback available, check to see if we have old style callbacks and use those instead (defaulting to SHA-256 for the hash).

This should be included in the beta.

This PR addresses most of the issues discussed in #5378. It does not attempt to resolve the issue that OpenSSL will (by default) prefer ciphersuites based on SHA-384 during ciphersuites selection, but old PSKs default to SHA-256 (and hence they may not be used).

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man3/SSL_CTX_set_psk_client_callback.pod [hunk 1]
```diff
@@ -113,6 +98,33 @@ case no PSK will be sent to the server but the handshake will continue. To do
...
+
+The purpose of the callback function is to select the PSK identity and
+the pre-shared key to use during the connection setup phase.
+
+The callback is set using functions SSL_CTX_set_psk_client_callback()
+or SSL_set_psk_client_callback(). The callback function is given the
+connection in parameter B<ssl>, a B<NULL>-terminated PSK identity hint
+sent by the server in parameter B<hint>, a buffer B<identity> of
+length B<max_identity_len> bytes where the resulting
+B<NULL>-terminated identity is to be stored, and a buffer B<psk> of
```
Comment (FdaSilvaYY) [inline doc/man3/SSL_CTX_set_psk_client_callback.pod]: NULL or NUL ?

**Comment by mattcaswell** [comment]
You meant NUL terminated right?

**Comment by richsalz** [comment]
Either way; NULL v NUL is hard to parse sometimes, zero-terminated is unambiguous.  Your choice.

**Comment by mattcaswell** [comment]
Pushed with nit fixed. Thanks.
